cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java

```diff
@@ -221,15 +222,22 @@ public void init(String[] args) {
   }
 
   @Override
-  public void start() {
+  public void start() throws Exception {
     logAppenderInitializer.initialize();
     createDirectory("twill");
     createDirectory(cConf.get(QueueConstants.ConfigKeys.QUEUE_TABLE_COPROCESSOR_DIR,
                               QueueConstants.DEFAULT_QUEUE_TABLE_COPROCESSOR_DIR));
     createSystemHBaseNamespace();
     updateConfigurationTable();
 
-    zkClient.startAndWait();
+    ListenableFuture<Service.State> startFunction = zkClient.start();
+    try {
+      startFunction.get(cConf.getLong(Constants.Zookeeper.CFG_CLIENT_TIMEOUT_MILLIS), TimeUnit.MILLISECONDS);
+    } catch (Exception e) {
```

Comment:
we should log a different message based on the type of exception. If its a TimeoutException, we want to log that it timed out and that the zk quorum setting should be checked for correctness.